PayloadPkg/FirmwareUpdate/FirmwareUpdate.c

```diff
@@ -511,6 +511,71 @@ EnforceFwUpdatePolicy (
   return EFI_SUCCESS;
 }
 
+/**
+  Reset state machine to processing.
+
+  This function resets the state machine to processing state.
+
+  @retval  EFI_SUCCESS      State machine reset occured successfully.
+  @retval  other            An error occured.
+**/
+EFI_STATUS
+ResetToProcessingState (
+  VOID
+  )
+{
+  EFI_STATUS            Status;
+  UINT32                FwUpdStatusOffset;
+  FW_UPDATE_STATUS      FwUpdStatus;
+  FW_UPDATE_COMP_STATUS FwUpdCompStatus[MAX_FW_COMPONENTS];
+
+  FwUpdStatusOffset = PcdGet32 (PcdFwUpdStatusBase);
+
+  // Read the current FW update status structure
+  Status = BootMediaRead (FwUpdStatusOffset, sizeof(FW_UPDATE_STATUS), (UINT8 *)&FwUpdStatus);
+  if (EFI_ERROR (Status)) {
+    DEBUG((DEBUG_ERROR, "BootMediaRead. offset: 0x%llx, Status = 0x%x\n", FwUpdStatusOffset, Status));
+    return Status;
+  }
+
+  // Read the current FW update component status structure array
+  Status = BootMediaRead ((FwUpdStatusOffset + sizeof(FW_UPDATE_STATUS)), \
+                           MAX_FW_COMPONENTS * sizeof(FW_UPDATE_COMP_STATUS), (UINT8 *)&FwUpdCompStatus);
+  if (EFI_ERROR (Status)) {
+    DEBUG((DEBUG_ERROR, "BootMediaRead. offset: 0x%llx, Status = 0x%x\n", (FwUpdStatusOffset + sizeof(FW_UPDATE_STATUS)), Status));
+    return Status;
+  }
+
+  // Change the SM to processing so that next redundant component update is triggered (if exists)
+  // Keep all other values the same
+  FwUpdStatus.StateMachine = FW_UPDATE_SM_CAP_PROCESSING;
+
+  // Need to erase because writes only bitwise-and the values
+  // Need to erase a whole page, because that's the smallest allowed
+  Status = BootMediaErase (FwUpdStatusOffset, EFI_PAGE_SIZE);
```

Comment:
This is tricky, if power failure happens, all the status saved will be lost. Need to consider a better solution

Reply:
how about consider individual state machine's in this case?

Reply:
Yeah, considering individual state machines works. I implemented this in the latest commit.